### PR TITLE
Subscriptions: Respect DisableAssetWebsocketSupport for assets

### LINKED ISSUE #1692: Subscriptions: Respect DisableAssetWebsocketSupport for assets
# New Issue

asset.Index is a unsupported websocket asset, however we don't respect this generating subs. We're also attempting to subscribe to the orderbook L2 for it which doesn't exist.

```
[DEBUG] | WEBSOCKET | 24/10/2024 15:40:04 | Bitmex wss://www.bitmex.com/realtimemd: Message received: [0,"public","public",{"status":400,"error":"Unknown or expired symbol: BBANANAT30M. Please see the documentation at https://www.bitmex.com/app/wsAPI.","meta":{},"request":{"op":"subscribe","args":["orderBookL2:APE_USDT","orderBookL2:BBANANAT30M","orderBookL2:ETHUSD","orderBookL2:XBTV24","trade:APE_USDT","trade:BBANANAT30M","trade:ETHUSD","trade:XBTV24"]}}]
[DEBUG] | WEBSOCKET | 24/10/2024 15:40:04 | Bitmex wss://www.bitmex.com/realtimemd: Message received: [0,"public","public",{"status":400,"error":"Unknown or expired symbol: BBANANAT30M. Please see the documentation at https://www.bitmex.com/app/wsAPI.","meta":{},"request":{"op":"subscribe","args":["orderBookL2:APE_USDT","orderBookL2:BBANANAT30M","orderBookL2:ETHUSD","orderBookL2:XBTV24","trade:APE_USDT","trade:BBANANAT30M","trade:ETHUSD","trade:XBTV24"]}}]
```

Looks like we need to add a period before the symbol for it to sub successfully to channels.

### PR DESCRIPTION
**Pull request by gbjk**
Subscriptions:
* Fix unsupported WS assets included

Bitmex:
* Assertify and fix tests
* Add index asset to test config

Fixes #1692 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] `TestGenerateSubscriptions` covers the requirement, once index was added to configtest

### CODE REVIEW

**Review by shazbert — CHANGES_REQUESTED** [review]
Minor thing, tested looks good 👍

File: exchanges/bitmex/bitmex_test.go [hunk 1]
```diff
@@ -635,10 +526,11 @@ func TestSubmitOrder(t *testing.T) {
...
 	response, err := b.SubmitOrder(context.Background(), orderSubmission)
-	if sharedtestvalues.AreAPICredentialsSet(b) && (err != nil || response.Status != order.New) {
-		t.Errorf("Order failed to be placed: %v", err)
-	} else if !sharedtestvalues.AreAPICredentialsSet(b) && err == nil {
-		t.Error("Expecting an error when no keys are set")
+	if sharedtestvalues.AreAPICredentialsSet(b) {
+		require.NoError(t, err)
+		assert.Equal(t, order.New, response.Status)
+	} else {
+		require.ErrorIs(t, err, exchange.ErrCredentialsAreEmpty)
```
[the remark this answers is not in the document]
Comment (gbjk) [reply exchanges/bitmex/bitmex_test.go:533]: The function won't be skipped if credentials are not set. See `SkipTestIfCannotManipulateOrders`.
The else condition is run.

I can change it so we call `sharedtestvalues.SkipTestIfCredentialsUnset` instead, and lose coverage on checking the error return. I thought about doing that but backed away from it.

[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/bitmex/bitmex_test.go:533]: whoopsie daisies, you are correct leave this alone.

**Review by shazbert — APPROVED** [review]
Approved

**Comment by gbjk** [comment]
Rebased

File: exchanges/gateio/gateio_test.go [hunk 2]
```diff
@@ -2972,7 +2972,7 @@ func TestGenerateSubscriptionsSpot(t *testing.T) {
 	exp := subscription.List{}
 	for _, s := range g.Features.Subscriptions {
 		for _, a := range g.GetAssetTypes(true) {
-			if s.Asset != asset.All && s.Asset != a {
+			if s.Asset != asset.All && s.Asset != a || !g.IsAssetWebsocketSupported(a) {
```
Comment (gbjk) [inline exchanges/gateio/gateio_test.go:2975]: This is not pretty.
Make it pretty.

Comment (gbjk) [reply exchanges/gateio/gateio_test.go]: Fixed https://github.com/thrasher-corp/gocryptotrader/pull/1693/commits/feed889b1e9f4b7124b20173bdd26fbd150f7a03

File: exchanges/subscription/list.go [hunk 3]
```diff
@@ -91,7 +92,12 @@ func fillAssetPairs(ap assetPairs, a asset.Item, e iExchange) error {
 
 // assetPairs returns a map of enabled pairs for the subscriptions in the list, formatted for the asset
 func (l List) assetPairs(e iExchange) (assetPairs, error) {
-	at := e.GetAssetTypes(true)
+	at := []asset.Item{}
+	for _, a := range e.GetAssetTypes(true) {
+		if e.IsAssetWebsocketSupported(a) {
```
Comment (gloriousCode) [inline exchanges/subscription/list.go:97]: With only `Index` asset enabled
Only trade sub enabled:
```
 {
      "enabled": true,
      "channel": "trade",
      "asset": "all"
     },
```

```
[ERROR] | WEBSOCKET | 14/11/2024 11:49:43 | Bitmex websocket: subscription failure, trade all : subscription template did not generate the expected number of asset records: Got 1; Expected 0
```

Do you classify this as something to address?

Comment (gbjk) [reply exchanges/subscription/list.go:97]: Absolutely. That's a bug in the sub template (which I've made a few times now).
I'll fix.

Comment (gbjk) [reply exchanges/subscription/list.go:97]: Note:
`Bitmex asset type index websocket functionality is unsupported, REST fetching only.`

**Review by gloriousCode — APPROVED** [review]
tACK!

**Review by shazbert — APPROVED** [review]
tACK! Thank you 🖖

**Comment by gbjk** [comment]
@gloriousCode I believe you're done, so I've added `gcrc` so this labels pretty for @thrasher- 🙂

**Review by thrasher- — APPROVED** [review]
tACK, thanks for resolving this issue!

**Pull request merged**